src/modules/musig/tests_impl.h

```diff
@@ -921,6 +923,367 @@ void musig_tweak_test(secp256k1_scratch_space *scratch) {
     }
 }
 
+int musig_vectors_keyagg_and_tweak(secp256k1_musig_keyagg_cache *keyagg_cache,
+                                   unsigned char *agg_pk_ser,
+                                   unsigned char pubkeys33[][33],
+                                   unsigned char tweaks32[][32],
+                                   size_t key_indices_len,
+                                   size_t *key_indices,
+                                   size_t tweak_indices_len,
+                                   size_t *tweak_indices,
+                                   int *is_xonly) {
+    secp256k1_pubkey pubkeys[MUSIG_VECTORS_MAX_PUBKEYS];
+    const secp256k1_pubkey *pk_ptr[MUSIG_VECTORS_MAX_PUBKEYS];
+    int i;
+    secp256k1_pubkey agg_pk;
+    secp256k1_xonly_pubkey agg_pk_xonly;
+
+    for (i = 0; i < (int)key_indices_len; i++) {
+        if (!secp256k1_ec_pubkey_parse(ctx, &pubkeys[i], pubkeys33[key_indices[i]], 33)) {
+            return 0;
+        }
+        pk_ptr[i] = &pubkeys[i];
+
+    }
+    if (!secp256k1_musig_pubkey_agg(ctx, NULL, NULL, keyagg_cache, pk_ptr, key_indices_len)) {
+        return 0;
+    }
+
+    for (i = 0; i < (int)tweak_indices_len; i++) {
+        if (is_xonly[i]) {
+            if (!secp256k1_musig_pubkey_xonly_tweak_add(ctx, NULL, keyagg_cache, tweaks32[tweak_indices[i]])) {
+                return 0;
+            }
+        } else {
+            if (!secp256k1_musig_pubkey_ec_tweak_add(ctx, NULL, keyagg_cache, tweaks32[tweak_indices[i]])) {
+                return 0;
+            }
+        }
+    }
+    if (!secp256k1_musig_pubkey_get(ctx, &agg_pk, keyagg_cache)) {
+        return 0;
+    }
+
+    if (!secp256k1_xonly_pubkey_from_pubkey(ctx, &agg_pk_xonly, NULL, &agg_pk)) {
+        return 0;
+    }
+
+    if (agg_pk_ser != NULL) {
+        if (!secp256k1_xonly_pubkey_serialize(ctx, agg_pk_ser, &agg_pk_xonly)) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+void musig_test_vectors_keyagg(void) {
+    int i;
+    struct musig_key_agg_vector vector;
+
+    create_key_agg_vector(&vector);
+
+    for (i = 0; i < (int)(sizeof(vector.valid_case)/sizeof(vector.valid_case[0])); i++) {
+        secp256k1_musig_keyagg_cache keyagg_cache;
+        unsigned char agg_pk[32];
+        struct musig_key_agg_valid_test_case *c = &vector.valid_case[i];
+
+        CHECK(musig_vectors_keyagg_and_tweak(&keyagg_cache, agg_pk, vector.pubkeys, vector.tweaks, c->key_indices_len, c->key_indices, 0, NULL, NULL));
+        CHECK(memcmp(agg_pk, c->expected, sizeof(agg_pk)) == 0);
```

Comment:
secp256k1_memcmp_var (also in a few other places)

Reply:
Would it make sense to allow regular `memcmp` in the tests?

Reply:
Maybe, but that's something that should be discussed upstream. I think if we believe that `memcmp` risky, then we should avoid it everywhere. But I know, it's very annoying, and I feel bad because I point this out in some many reviews. I guess we should have a check in CI...

Reply:
I do think it makes a significant difference whether memcmp is used in tests or not. The only risk I can see with a memcmp like the one above is that miscompilation may hide an actual bug. But the miscompilation issue has been fixed since in GCC and at least you and I and apoelstra don't use the buggy GCC versions. Also CI runs almost all of the tests with also with clang.

Reply:
Yep, valid points. I'm not going to insist on this obviously, so either is fine here for now, but let's maybe restart this discussing in https://github.com/bitcoin-core/secp256k1/issues/823?

Reply:
Yep, valid points. I'm not going to insist on this obviously, so either is fine here for now, but let's maybe restart this discussing in https://github.com/bitcoin-core/secp256k1/issues/823?

Reply:
Yep, valid points. I'm not going to insist on this obviously, so either is fine here for now, but let's maybe restart this discussing in https://github.com/bitcoin-core/secp256k1/issues/823?

Reply:
changed memcmp to secp256k1_memcmp_var